index.bs

```diff
@@ -1810,7 +1809,71 @@ partial interface MLGraphBuilder {
 </div>
 
 <details open algorithm>
+  <summary>
+    To <dfn for=MLGraphBuilder>calculate conv2d padding</dfn> given unsigned integers |inputSize|, |filterSize|, |stride| and |dilation| and {{MLAutoPad}} |autoPad|, perform these steps. They return a [=/list=] of 2 numbers.
+  </summary>
+  <div class=algorithm-steps>
+    1. [=Assert=]: |autoPad| is not {{MLAutoPad/"explicit"}}.
+    1. Let |outputSize| be ( |inputSize| + |stride| - 1 ) / |stride|.
+    1. Let |dilatedFilterSize| be ( |filterSize| - 1 ) * |dilation| + 1.
+    1. Let |neededInputSize| be ( |outputSize| - 1 ) * |stride| + |dilatedFilterSize|.
+    1. Let |totalPadding| be |neededInputSize| - |inputSize| if |neededInputSize| is greater than |inputSize|, or 0 otherwise.
+    1. Switch on |autoPad|:
+        <dl class=switch>
+            : {{MLAutoPad/"same-upper"}}
+            ::
+                1. Let |paddingBegin| be floor( |totalPadding| / 2 ).
+                1. Let |paddingEnd| be floor( ( |totalPadding| + 1 ) / 2 ).
+            : {{MLAutoPad/"same-lower"}}
+            ::
+                1. Let |paddingBegin| be floor( ( |totalPadding| + 1 ) / 2 ).
+                1. Let |paddingEnd| be floor( |totalPadding| / 2 ).
+        </dl>
+    1. Return « |paddingBegin|, |paddingEnd| ».
+  </div>
+</details>
+
+
+<details open algorithm>
+  <summary>
+    To <dfn for=MLGraphBuilder>calculate conv2d output size</dfn> given unsigned integers |inputSize|, |filterSize|, |beginningPadding|, |endingPadding|, |stride| and |dilation|, perform these steps. They return a number.
+  </summary>
+  <div class=algorithm-steps>
+    1. Let |effectiveFilterSize| be ( |filterSize| - 1 ) * |dilation| + 1.
+    1. Let |outputSize| be ( |inputSize| - |effectiveFilterSize| + |beginningPadding| + |endingPadding| ) / |stride| + 1.
```

Comment:
Should we handle any errors? Like |outputSize| is overflow or underflow?

Reply:
That's a good question... the specs I've been involved in usually operate in an abstract space with infinite precision, unlimited range, etc., with very limited checking. 

I noticed the Chromium prototype impl is using checked math (i.e. testing underflow/overflow), and given the sizes that ML deals with it, it seems like this is going to be a practical concern. 

We can spec this any way we want... ideas welcome. We should look at other specs for inspiration, too.

Reply:
Hmm, I've always found it cleaner when implementing specs (like bidi, line breaking, vertical layout...) that I was grateful they left out that degree of fine validation and instead focused on the algorithm itself. Any validation specific to the nature of the operators (e.g. maybe it only supports even sizes) belongs in there, but too much otherwise muddies an already complex document. Maybe we can even call that out somewhere in a common section, that the implementation should handle overflow/underflow/safe limit checks, but not as extra prose for each operator.